### ISSUE TITLE
Feature request: Async field level validation

### ISSUE THREAD

**Issue by lloydh** [erikras/redux-form#2763]
It would be great to be able to use promise-based validation libraries like yup for field level validation but it seems that redux-form only supports sync validation functions on fields.

There's a [stack overflow post](http://stackoverflow.com/questions/40363850/is-it-bad-practice-to-repurpose-asyncvalidate-for-integrating-a-promise-returnin) and [this article](http://www.davidboyne.co.uk/2016/04/29/redux-form-validation-with-yup.html) discussing use of the asyncValidate config option but I don't think there are any options for async field level validation.

**Comment by jorrit** [comment]
I would like this feature as well. It is a bit cumbersome to set the async function and asyncBlurFields on the form level. It would be nicer to set asyncValidate on a field and get the same effect, just like with synchronous validation.

**Comment by roytz** [comment]
Will be very useful to me too! Will this feature be available anytime soon?
Thanks!

**Comment by jorrit** [comment]
Perhaps a maintainer can make an estimate of the effort needed. If it's not too much I could take a look.

**Comment by shai32** [comment]
thats will be great

**Comment by yussinsharp** [comment]
Wrote some code that allowed me to compose asyncValidator functions into one. Thought I'd share here:

```js
function composeAsyncValidators(validatorFns) {
  return async (values, dispatch, props, field) => {
    let errors = {};
    for (const validatorFn of validatorFns) {
      try {
        await validatorFn(values, dispatch, props, field);
      } catch (err) {
        errors = Object.assign({}, errors, err);
      }
    }

    throw errors;
  };
}

// Example Usage - need to call multiple API endpoints to validate data

function uniqueEmail() {
  return axios.post('/users/phoneNumber', { phoneNumber: values.phoneNumber })
    .then((response) => {
      if (response.data.exists) {
        throw { phoneNumber: 'There is already a user with that phone number' };
      }
    });
}

function uniquePhoneNumber() {
  return axios.post('/users/phoneNumber', { phoneNumber: values.phoneNumber })
    .then((response) => {
      if (response.data.exists) {
        throw { phoneNumber: 'There is already a user with that phone number' };
      }
    });
}

const asyncValidate = composeAsyncValidators([uniqueEmail, uniquePhoneNumber]);
```

Field-level async validation would've been preferred but this also solves the case of composing functions that depend on multiple fields (or throw errors for multiple fields).

**Comment by jorrit** [comment]
Isn't something involving `Promise.all()` appropriate here?

**Comment by bitstrider** [comment]
@yussinsharp this looks like a good start for a workaround, but it will call all async validation fn when blurring a single field. And then there's the issue with how to render the errors for the specific field.

To be sure, here's an example:
https://codesandbox.io/s/L9Y53R8w4

**Comment by yussinsharp** [comment]
Yes it will. I wrote it like that in case a single asyncValidation depended on multiple fields and also because the thrown errors will do a hard reset of the errors on the form. So if we coded it in a way to not call an async function if it doesn't have to, we'd have to remember to get the previously thrown errors added in as well. Meaning you'd have to go deeper into the inner workings of redux-form and find new ways to manage its state.

**Comment by bitstrider** [comment]
Yeah, we can map the validation functions to the fields as such:
```javascript
function composeAsyncValidators(validatorFns) {
  return async (values, dispatch, props, field) => {
    const validatorFn = validatorFns[field]
    await validatorFn(values, dispatch, props, field);
  };
}

/* ... */

const asyncValidate = composeAsyncValidators({
  username:usernameValidate,
  companyname:companynameValidate
});
```

But still only the last async error gets presented:
https://codesandbox.io/s/6R4xMwXOQ

From debugging the redux state, it looks like under-the-hood only the last async error gets stored to the `asyncErrors` state slice:

![redux-multiple-async](https://user-images.githubusercontent.com/143295/27712564-9d05e7c6-5cf5-11e7-873f-33538c2fbf4f.png)

**Comment by joanoso** [comment]
Is someone going to make this async field value validation. If not, i will have to change to another library.

**Comment by reisel** [comment]
@joanoso Please don't!

**Comment by sckoh** [comment]
@yussinsharp @bitstrider you can merge the previous asyncFormErrors by putting previous asyncErrors into your props

```
const mapStateToProps = (state, props) => ({
    formAsyncErrors: getFormAsyncErrors(FORM_NAME)(state),
})


const asyncValidate = (
  values,
  dispatch,
  props,
  asyncValidatingField
) => {
  const previousErrors = { ...props.formAsyncErrors };
  // then you can continue to use @yussinsharp way to catch each validator function error
  // and merge it with previousErrors

```

**Comment by Ugzuzg** [comment]
@sckoh, `asyncErrors` field is already in the props, there's no need to map it manually. The only downside to this method is if you fire async validation for a different field and the validation for the first field isn't finished yet, you'll override the result of one of the validations.

**Comment by sckoh** [comment]
@Ugzuzg I found that asyncErrors in the props is not updated, I always need to map it myself.
You dont need to override it. You can use a single errors object to store all the field asyncErrors

**Comment by damianobarbati** [comment]
+1 for this feature, async field validation would be a very plus for this project

**Comment by lloydh** [comment]
I hope to see async field validation come to redux-form but it's worth noting that the validation library yup (which prompted this feature request) has just released sync validation support.

**Comment by damianobarbati** [comment]
@erikras any chance to see this in the near future?

**Comment by jsonmaur** [comment]
Seems like the `asyncValidate` function only occurs on blur. This feature seems necessary if you want to use async validation as the user types (debounced) instead of on blur. Is there another way to accomplish this with the current api?

**Comment by shepmaster** [comment]
@jsonmaur that would be #3521, I believe.

**Comment by damianobarbati** [comment]
yes but we are talking about "workarounds": we are trying to use existing functionalities to achieve another desired functionality which should be implemented in a dedicated/ad-hoc way.
Actually specifying "blur fields" array or handler is not as clean as having an `asyncValidators` prop on the wanted field (imho).

**Comment by lvian** [comment]
Is this going forward?

I was looking forward to this functionality for something rather simple, I wan't to  validate the dimensions of my images before submitting, but since the Image onLoad function is async I'm stuck.

Hope this happens in the future

**Comment by gustavohenke** [comment]
Hi @lvian, unfortunately we're not seeing many new implementations coming through.
If someone has spare time to work on this, that would be great.

**Comment by Gtosta96** [comment]
@erikras it would be very useful...

**Comment by odrzutowiec** [comment]
+1 I need this! How come it's not there?! Seems like a super major important feature!

**Comment by DanyST** [comment]
@odrzutowiec works fine!
Thanks!

**Comment by async3619** [comment]
I wrote some of HoC to solve this problem but it still seems not the good way to solve since it looks decreasing performance so much.

I know that keeping asking about this is rude but do any of contributors mind to implement async field-level validation?

**Comment by Bram-Zijp** [comment]
The example of @odrzutowiec  did not work for me. It would label invalid fields valid. 

This works for me;

```
const asyncValidate = (values, dispatch, meta, field) =>
  new Promise(async (resolve, reject) => {
    let email = {}
    let username = {}
    if (field === "email") {
      email = await emailAsyncValidate(values.email)
    }
    if (field === "username") {
      username = await usernameAsyncValidate(values.username)
    }
    const result = { ...meta.asyncErrors, ...email, ...username, }
    if (!Object.keys(result).length) resolve()
    reject(result)
  })
```
